tools/replica/reorder_remd_traj.py

```diff
@@ -0,0 +1,571 @@
+#!/usr/bin/env python
+
+"""
+LAMMPS Replica Exchange Molecular Dynamics (REMD) trajectories are arranged by
+replica, i.e., each trajectory is a continuous replica that records all the 
+ups and downs in temperature. However, often the requirement is trajectories
+that are continuous in temperature, which is achieved by this tool.
+
+Author: 
+Tanmoy Sanyal, Shell lab, Chemical Engineering, UC Santa Barbara
+Email: tanmoy dot 7989 at gmail dot com
+
+Usage
+-----
+To get detailed information about the arguments, flags, etc use:
+python reorder_remd_traj.py -h or 
+python reorder_remd_traj.py --help
+
+Features of this script
+-----------------------
+a) reorder LAMMPS REMD trajectories by temperature keeping only desired frames.
+Note: this only handles LAMMPS format trajectories (i.e. .lammpstrj format)
+Trajectories can be gzipped or bz2-compressed. The trajectories are assumed to
+be named as <prefix>.%d.lammpstrj[.gz or .bz2]
+
+b) (optionally) calculate configurational weights for each frame at each
+temperature if potential energies are supplied. But this if for the canonical
+(NVT) ensemble only.
+
+Dependencies
+------------
+mpi4py
+pymbar (for getting configurational weights)
+tqdm (for printing pretty progress bars)
+StringIO (or io if in Python 3.x)
+
+"""
+
+
+
+import os, sys, numpy as np, argparse, time, pickle
+from scipy.special import logsumexp
+from mpi4py import MPI
+
+from tqdm import tqdm, trange
+import gzip
+try:
+    # python-2
+    from StringIO import StringIO as IOBuffer
+except ImportError:
+    # python-3
+    from io import BytesIO as IOBuffer
+
+
+
+#### INITIALISE MPI ####
+# (note that all output on screen will be printed only on the ROOT proc)
+ROOT = 0
+comm = MPI.COMM_WORLD
+me = comm.rank # my proc id
+nproc = comm.size
+
+
+#### HELPER FUNCTIONS ####
+def _get_nearest_temp(temps, query_temp):
+    """
+    Helper function to get the nearest temp in a list
+    from a given query_temp
+    
+    :param temps: list of temps.
+    
+    :param query_temp: query temp
+    
+    Returns:
+    idx: index of nearest temp in the list
+    
+    out_temp: nearest temp from the list
+    """
+    
+    if isinstance(temps, list): temps = np.array(temps)
+    idx = np.argmin(abs(temps - query_temp))
+    out_temp = temps[idx]
+    return out_temp
+
+
+def readwrite(trajfn, mode = "rb"):
+    """ 
+    Helper function for input/output LAMMPS traj files.
+    Trajectories may be plain text, .gz or .bz2 compressed.
+        
+    :param trajfn: name of LAMMPS traj
+    
+    :param mode: "r" ("w") and "rb" ("wb") depending on read or write
+        
+    Returns: file pointer
+    """
+    
+    if trajfn.endswith(".gz"):
+        return gzip.GzipFile(trajfn, mode)
+    elif trajfn.endswith(".bz2"):
+        return bz2.BZ2File(trajfn, mode)
+    else:
+        return file(trajfn, mode)
+
```

Comment:
I do not think there is a file function in python. The trajfn file should be opened with the open() function. Same holds true for the gzip and bz2 cases

Reply:
agreed. changed all the methods to their corresponding open methods. file() is a legacy method that somehow works my anaconda distribution, but its not recommended anymore. Thanks for pointing this out!